### PR TITLE
Gracefully handle postgres pool timeouts

### LINKED ISSUE #626: Panic/timeout connecting to postgres
```
Feb 25 19:38:47 docsrs cratesfyi[29874]: thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Error(None)', src/libcore/result.rs:1165:5
Feb 25 19:38:47 docsrs cratesfyi[29874]: stack backtrace:
Feb 25 19:38:47 docsrs cratesfyi[29874]: thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Error(None)', src/libcore/result.rs:1165:5
Feb 25 19:38:47 docsrs cratesfyi[29874]: thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Error(None)', src/libcore/result.rs:1165:5
Feb 25 19:38:47 docsrs cratesfyi[29874]: thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Error(None)', src/libcore/result.rs:1165:5
...
Feb 25 19:38:48 docsrs cratesfyi[29874]:   14: core::result::unwrap_failed
Feb 25 19:38:48 docsrs cratesfyi[29874]:              at src/libcore/result.rs:1165
Feb 25 19:38:48 docsrs cratesfyi[29874]:   15: <cratesfyi::web::pool::Pool as iron::middleware::BeforeMiddleware>::before
Feb 25 19:38:48 docsrs cratesfyi[29874]:   16: iron::middleware::Chain::continue_from_before
Feb 25 19:38:48 docsrs cratesfyi[29874]:   17: <iron::middleware::Chain as iron::middleware::Handler>::handle
Feb 25 19:38:48 docsrs cratesfyi[29874]:   18: <alloc::boxed::Box<dyn iron::middleware::Handler> as iron::middleware::Handler>::handle
Feb 25 19:38:48 docsrs cratesfyi[29874]:   19: <cratesfyi::web::CratesfyiHandler as iron::middleware::Handler>::handle
Feb 25 19:38:48 docsrs cratesfyi[29874]:   20: <iron::iron::RawHandler<H> as hyper::server::Handler>::handle
Feb 25 19:38:48 docsrs cratesfyi[29874]:   21: hyper::server::Worker<H>::handle_connection
Feb 25 19:38:48 docsrs cratesfyi[29874]:   22: hyper::server::listener::spawn_with::{{closure}}
Feb 25 19:38:48 docsrs cratesfyi[29874]: note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
...
Feb 25 19:40:34 docsrs cratesfyi[29874]: thread 'thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Error(None)', src/libcore/result.rs:<unnamed>' 
Feb 25 19:40:34 docsrs cratesfyi[29874]: 1165:5
Feb 25 19:40:35 docsrs kernel: nginx invoked oom-killer: gfp_mask=0x14200ca(GFP_HIGHUSER_MOVABLE), nodemask=(null), order=0, oom_score_adj=0
```

Relevant code: https://github.com/rust-lang/docs.rs/blob/master/src/web/pool.rs#L34

### PR DESCRIPTION
**Pull request by Kixiron**
Closes #626

* Changes from unwrapping on a failure to get a pool to returning an error
* Adds the `docsrs_failed_db_connections` metric to monitor timed out database connections
  * A metric for concurrent database connections could also be added

### CODE REVIEW

File: src/web/pool.rs [hunk 1]
```diff
@@ -1,10 +1,12 @@
...
-use iron::{typemap, BeforeMiddleware, IronResult, Request};
+use iron::{status::Status, typemap, BeforeMiddleware, IronError, IronResult, Request};
 use postgres::Connection;
+use std::marker::PhantomData;
 
 #[cfg(test)]
 use std::sync::{Arc, Mutex, MutexGuard};
 
+#[derive(Debug, Clone)]
 pub(crate) enum Pool {
```
Comment (jyn514) [inline src/web/pool.rs:10]: Won't this `Clone` still create more pools? I thought the idea was to limit the number of concurrent connections by using a single pool that kept track of them itself.

**Comment by jyn514** [comment]
How did you test these changes?

File: src/web/pool.rs (hunk 1 above)
Comment (Kixiron) [reply src/web/pool.rs:10]: `Pool<T>` is just an [`Arc`](https://docs.rs/r2d2/0.8.8/src/r2d2/lib.rs.html#317), so it clones that

**Comment by Kixiron** [comment]
Testing was done via Apache's `ab`, if the server's stressed under enough concurrent load then errors start popping up.

Does bring up the issue that the scoping of connections needs improvement, but that's another PR

File: src/db/mod.rs [hunk 2]
```diff
@@ -28,10 +28,13 @@ pub fn connect_db() -> Result<Connection, failure::Error> {
 pub(crate) fn create_pool() -> r2d2::Pool<r2d2_postgres::PostgresConnectionManager> {
     let db_url = env::var("CRATESFYI_DATABASE_URL")
         .expect("CRATESFYI_DATABASE_URL environment variable is not exists");
+
     let manager =
         r2d2_postgres::PostgresConnectionManager::new(&db_url[..], r2d2_postgres::TlsMode::None)
             .expect("Failed to create PostgresConnectionManager");
+
     r2d2::Pool::builder()
+        .max_size(90)
```
Comment (pietroalbini) [inline src/db/mod.rs:37]: Let's make this configurable with an environment variable.

**Comment by pietroalbini** [comment]
> A metric for concurrent database connections could also be added

This would be really useful to determine how close we usually are to capacity btw!

**Comment by pietroalbini** [comment]
Do you plan to add the pool capacity metric in this PR or in a followup?

**Comment by Kixiron** [comment]
Sorry, I've just been a bit busy

**Comment by pietroalbini** [comment]
Thanks :heart:

**Pull request merged**